timemachine/cpp/src/potential.cu

```diff
@@ -1,16 +1,112 @@
-#include <iostream>
 #include <memory>
 
 #include "device_buffer.hpp"
 #include "fixed_point.hpp"
 #include "gpu_utils.cuh"
 #include "potential.hpp"
-#include "surreal.cuh"
+
+#include <chrono>
+#include <iostream>
 
 namespace timemachine {
 
 const int Potential::D = 3;
 
+void Potential::execute_batch_host(
+    const int coord_batches,     // Number of batches of coordinates
+    const int N,                 // Number of atoms
+    const int param_batches,     // Number of batches of parameters
+    const int P,                 // Number of parameters
+    const int lambda_batches,    // Number of lambda values
+    const double *h_x,           // [coord_batches, N, 3]
+    const double *h_p,           // [param_batches, P]
+    const double *h_box,         // [coord_batches, 3, 3]
+    const double *lambdas,       // [lambda_batches]
+    unsigned long long *h_du_dx, // [coord_batches, param_batches, lambda_batches, N, 3]
+    unsigned long long *h_du_dp, // [coord_batches, param_batches, lambda_batches, P]
+    unsigned long long *h_du_dl, // [coord_batches, param_batches, lambda_batches, N]
+    unsigned long long *h_u) {   // [coord_batches, param_batches, lambda_batches, N]
```

Comment:
Nit: naming, would `batch_size` be more accurate than `batches`?

(Related curiosity: if these imply a result that is too large to be stored on the device at once, I wonder if this (or perhaps better, the `execute_batch_host` function) should split the job into multiple smaller batches...)

Reply:
Will rename to `*_batch_size`

Right now if you set a result that is too large, this will crash. We could try and track memory usage in the future, though that is usually a fraught task (at least with CPU memory).